src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+                if (!this.chainedHeadersByHash.ContainsKey(currentBlockHash))
+                {
+                    this.logger.LogTrace("A new header with hash '{0}' was found that is not connected to the tree.", currentBlockHash);
+                    this.logger.LogTrace("(-):true");
+                    return true;
+                }
+            }
+
+            this.logger.LogTrace("(-):false,{0}:{1}", nameof(newHeaderPosition), newHeaderPosition);
+            return false;
+        }
+
+        /// <summary>
+        /// Checks if <paramref name="chainedHeader"/> violates the max reorg rule, if networks.Consensus.MaxReorgLength is zero this logic is disabled.
+        /// </summary>
+        /// <param name="chainedHeader">The header that needs to be checked for reorg.</param>
+        /// <returns><c>true</c> if maximum reorg rule violated, <c>false</c> otherwise.</returns>
+        private void CheckMaxReorgRuleViolated(ChainedHeader chainedHeader)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeader), chainedHeader);
+
+            ChainedHeader tip = this.chainState.ConsensusTip.FindFork(chainedHeader);
+
+            uint maxReorgLength = this.chainState.MaxReorgLength;
+            ChainedHeader consensusTip = this.chainState.ConsensusTip;
+            if ((maxReorgLength != 0) && (consensusTip != null))
+            {
+                ChainedHeader fork = tip.FindFork(consensusTip);
+
+                if ((fork != null) && (fork != consensusTip))
+                {
+                    int reorgLength = consensusTip.Height - fork.Height;
+
+                    if (reorgLength > maxReorgLength)
+                    {
+                        this.logger.LogTrace("Reorganization of length {0} prevented, maximal reorganization length is {1}, consensus tip is '{2}'.", reorgLength, maxReorgLength, consensusTip);
+                        this.logger.LogTrace("(-)[MAX_REORG_VIOLATION]");
+                        throw new MaxReorgViolationException();
+                    }
+
+                    this.logger.LogTrace("Reorganization of length {0} accepted, consensus tip is '{1}'.", reorgLength, consensusTip);
+                }
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+    }
+
+    /// <summary>
+    /// Represents the response form the <see cref="ChainedHeaderTree.ConnectNewHeaders"/> method.
+    /// </summary>
+    public class ConnectedHeaders
+    {
+        /// <summary>The earliest header in the chain of the list of headers we are interested in downloading.</summary>
+        public ChainedHeader DownloadFrom { get; set; }
+
+        /// <summary>The latest header in the chain of the list of headers we are interested in downloading.</summary>
+        public ChainedHeader DownloadTo { get; set; }
+
+        /// <summary>Represent the chain of headers that are where added to the tree.</summary>
```

Comment:
It's more like a last consumed header from the ones that were announced by the peer.

Reply:
Represent`s`